Additional Brave Sync P3A metrics: object count and enabled types

AlexeyBarabash opened the pull request:
<!-- Add brave-browser issue below that this PR will resolve -->
Resolves https://github.com/brave/brave-browser/issues/29549

<!-- CI-related labels that can be applied to this PR:
* CI/run-audit-deps (1) - check for known npm/cargo vulnerabilities (audit_deps)
* CI/run-network-audit (1) - run network-audit
* CI/run-upstream-tests - run Chromium unit and browser tests on Linux and Windows (otherwise only on Linux)
* CI/skip - do not run CI builds (except noplatform)
* CI/skip-linux-arm64, CI/skip-linux-x64, CI/skip-android, CI/skip-macos, CI/skip-ios, CI/skip-windows-arm64, CI/skip-windows-x64, CI/skip-windows-x86 - skip CI builds for specific platforms
* CI/skip-upstream-tests - do not run Chromium unit, or browser tests (otherwise only on Linux)
* CI/storybook-url (1) - deploy storybook and provide a unique URL for each build

(1) applied automatically when some files are changed (see: https://github.com/brave/brave-core/blob/master/.github/labeler.yml)
-->

## Submitter Checklist:

- [x] I confirm that no security/privacy review [is needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or that I have [requested](https://github.com/brave/security/issues/new/choose) one
- I asked for review at our dedicated Slack channel and got the approval
- [x] There is a [ticket](https://github.com/brave/brave-browser/issues) for my issue
- [x] Used Github [auto-closing keywords](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue) in the PR description above
- [ ] Wrote a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- [x] Squashed any review feedback or "fixup" commits before merge, so that history is a record of what happened in the repo, not your PR
- [x] Added appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- [ ] Checked the PR locally:
  * `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests` [wiki](https://github.com/brave/brave-browser/wiki/Tests)
  * `npm run lint`, `npm run presubmit` [wiki](https://github.com/brave/brave-browser/wiki/Presubmit-checks), `npm run gn_check`, `npm run tslint`
- [x] Ran `git rebase master` (if needed)

## Reviewer Checklist:

- [ ] A security review [is not needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or a link to one is included in the PR description
- [ ] New files have MPL-2.0 license header
- [ ] Adequate test coverage exists to prevent regressions
- [ ] Major classes, functions and non-trivial code blocks are well-commented
- [ ] Changes in component dependencies are properly reflected in `gn`
- [ ] Code follows the [style guide](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md)
- [ ] Test plan is specified in PR before merging

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on
- [ ] All relevant documentation has been updated, for instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Custom-Headers
  - [ ] https://github.com/brave/brave-browser/wiki/Web-Compatibility-Exceptions-in-Brave
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide
  - [ ] https://github.com/brave/brave-browser/wiki/P3A

## Test Plan:
Ensure you able to see the metrics of `Brave.Sync.SyncedObjectsCount` and `Brave.Sync.EnabledTypes` with values at P3A dashboard

cdesouza-chromium commented on components/brave_sync/brave_sync_p3a.cc:
```diff
@@ -0,0 +1,65 @@
+  // 0 - no types or only Bookmarks
+  // 1 - Bookmarks + History
+  // 2 - more than (Bookmarks + History) but less than (Sync All)
+  // 3 - Sync everything
```
Have you considered making this an `enum class`. There's [UmaHistogramEnumeration](https://source.chromium.org/chromium/chromium/src/+/main:base/metrics/histogram_functions.h;bpv=1;bpt=1;l=71?q=UmaHistogramExactLinear&ss=chromium%2Fchromium%2Fsrc&gsn=UmaHistogramEnumeration&gs=KYTHE%3A%2F%2Fkythe%3A%2F%2Fchromium.googlesource.com%2Fcodesearch%2Fchromium%2Fsrc%2F%2Fmain%3Flang%3Dc%252B%252B%3Fpath%3Dbase%2Fmetrics%2Fhistogram_functions.h%23j7iyoHKtSe1QPv-Yb178Iy9gAKfQuXFdhQp-vqUmg-w).

If you think that might look better, make sure your enum has a `kMaxValue`, i.e. `enum class Foo {kFirst, kSecond, kThird, kMaxValue = kThird}`, that way you annotate what is the max value when calling `UmaHistogramEnumeration`

AlexeyBarabash replied:
Thanks, this notice helped make PR more clear: https://github.com/brave/brave-core/pull/17887/commits/06674f5a4260d7492a25afe2454e6f135f528ed9

cdesouza-chromium replied:
Ack.

cdesouza-chromium commented on components/sync/driver/brave_sync_service_impl.cc:
```diff
@@ -305,4 +307,40 @@ void BraveSyncServiceImpl::LocalDeviceAppeared() {
+void BraveSyncServiceImpl::OnSyncCycleCompleted(
+    const SyncCycleSnapshot& snapshot) {
+  SyncServiceImpl::OnSyncCycleCompleted(snapshot);
+  if (completed_cycles_count_ % kCyclesBeforeUpdateP3AObjects == 0) {
```
nit. I think it might be helpful to have a comment here explaining what is being checked in this conditional, as the modulo there seems a bit elusive to me.

cdesouza-chromium reviewed (approved):
Thanks for this PR.

I'm not very familiar with these changes, but overall LGTM.

cdesouza-chromium commented on components/sync/driver/brave_sync_service_impl.cc:
```diff
@@ -305,4 +307,40 @@ void BraveSyncServiceImpl::LocalDeviceAppeared() {
+void BraveSyncServiceImpl::OnSyncCycleCompleted(
+    const SyncCycleSnapshot& snapshot) {
+  SyncServiceImpl::OnSyncCycleCompleted(snapshot);
+  if (completed_cycles_count_ % kCyclesBeforeUpdateP3AObjects == 0) {
```
nit: `DCHECK_GT(completed_cycles_count_, 0)` might catch some unknown issue.

cdesouza-chromium replied:
Ah, so it doesn't matter. Acknowledged.

DJAndries reviewed (approved):
Pushed a commit to move metrics to be reporting on a monthly cadence as discussed, and fixed a typo in the metric name. everything else lgtm!

AlexeyBarabash commented:
Rebased and squashed. Current master fails on https://github.com/brave/brave-browser/issues/29571 , this PR CI will also fails. I'll have to rebase it again.
